rgrow/src/models/sdc1d.rs

```diff
@@ -0,0 +1,805 @@
+    /// Each strand will be given a color so that it can be easily identified
+    /// when illustrated
+    pub colors: Vec<[u8; 4]>,
+    /// The (de)attachment rates will depend on this constant(for the system) value
+    pub kf: RatePerConc,
+    /// FIXME: Change this to a vector to avoid hashing time
+    ///
+    /// Set of tiles that can stick to scaffold gap with a given glue
+    pub friends_btm: HashMap<Glue, HashSet<Tile>>,
+    /// Delta G at 37 degrees C in the formula to genereate the glue strengths
+    pub delta_g_matrix: Array2<f64>,
+    /// S in the formula to geenrate the glue strengths
+    pub entropy_matrix: Array2<f64>,
+    /// Temperature of the system
+    ///
+    /// Not pub so that it cant accidentally be changed other than with the setter function
+    /// that will also recalculate energy arrays
+    temperature: f64,
+    /// The energy with which two strands will bond
+    ///
+    /// This array is indexed as follows. Given strands x and y, where x is to the west of y
+    /// (meaning that the east of x forms a bond with the west of y), the energy of said bond
+    /// is given by energy_bonds[(x, y)]
+    strand_energy_bonds: Array2<Energy>,
+    /// The energy with which a strand attached to scaffold
+    scaffold_energy_bonds: Array1<Energy>,
+    /// Binding strength between two glues
+    glue_links: Array2<Strength>,
+}
+
+impl SDC {
+    fn new(
+        anchor_tiles: Vec<(PointSafe2, Tile)>,
+        strand_names: Vec<String>,
+        glue_names: Vec<String>,
+        scaffold: Array2<Glue>,
+        strand_concentration: Array1<Conc>,
+        glues: Array2<Glue>,
+        colors: Vec<[u8; 4]>,
+        kf: RatePerConc,
+        delta_g_matrix: Array2<f64>,
+        entropy_matrix: Array2<f64>,
+        temperature: f64,
+    ) -> SDC {
+        let strand_count = strand_names.len();
+        let mut s = SDC {
+            anchor_tiles,
+            strand_concentration,
+            strand_names,
+            colors,
+            glues,
+            scaffold,
+            glue_names,
+            kf,
+            delta_g_matrix,
+            entropy_matrix,
+            temperature,
+            // These will be generated by the update_system function next, so just leave them
+            // empty for now
+            friends_btm: HashMap::new(),
+            glue_links: Array2::<f64>::zeros((strand_count, strand_count)),
+            strand_energy_bonds: Array2::<f64>::zeros((strand_count, strand_count)),
+            scaffold_energy_bonds: Array1::<f64>::zeros(strand_count),
+        };
+        s.update_system();
+        s
+    }
+
+    fn update_system(&mut self) {
+        // Note that order is important, we need to generate the glue matrix first, then using
+        // the data generated there, the energy array is filled, etc...
+        self.generate_glue_matrix();
+        self.fill_energy_array();
+        self.generate_friends();
+    }
+
+    fn generate_friends(&mut self) {
```

Comment:
The way this is implemented is for matching glues, but doesn't work for complementary glues, eg, "a" an tile and "a*" on scaffold, resulting in glue number 5 on tile and 6 on scaffold, with glue_links[(5,6)] < 0.